android/libraries/rib-base/src/main/java/com/badoo/ribs/core/routing/transition/effect/SharedElementTransition.kt

```diff
@@ -0,0 +1,169 @@
+@file:SuppressWarnings("LongMethod")
+package com.badoo.ribs.core.routing.transition.effect
+
+import android.animation.Animator
+import android.animation.AnimatorListenerAdapter
+import android.animation.ValueAnimator
+import android.view.View
+import android.view.ViewGroup
+import android.view.animation.Interpolator
+import android.view.animation.LinearInterpolator
+import androidx.annotation.CheckResult
+import com.badoo.ribs.core.routing.transition.Transition
+import com.badoo.ribs.core.routing.transition.TransitionDirection
+import com.badoo.ribs.core.routing.transition.TransitionElement
+import com.badoo.ribs.core.routing.transition.handler.defaultDuration
+import com.badoo.ribs.core.routing.transition.handler.defaultInterpolator
+import com.badoo.ribs.core.routing.transition.progress.SingleProgressEvaluator
+import com.badoo.ribs.core.routing.transition.effect.SharedElementTransition.RotationParams
+
+interface SharedElementTransition {
+    data class Params(
+        val duration: Long = defaultDuration,
+        val exitingElement: (View) -> View?,
+        val enteringElement: (View) -> View?,
+        val translateXInterpolator: Interpolator = LinearInterpolator(),
+        val translateYInterpolator: Interpolator = LinearInterpolator(),
+        val scaleXInterpolator: Interpolator = LinearInterpolator(),
+        val scaleYInterpolator: Interpolator = LinearInterpolator(),
+        val rotation: RotationParams? = null,
+        val rotationX: RotationParams? = null,
+        val rotationY: RotationParams? = null
+    )
+
+    data class RotationParams(
+        val degrees: Float,
+        val interpolator: Interpolator = defaultInterpolator
+    )
+}
+
+internal data class SharedElementTransitionInfo<T>(
+    val exitingElement: TransitionElement<out T>,
+    val exitingView: View,
+    val enteringElement: TransitionElement<out T>,
+    val enteringView: View,
+    val params: SharedElementTransition.Params
+)
+
+@CheckResult
+fun <T> List<TransitionElement<out T>>.sharedElementTransition(
+    transitionParams: List<SharedElementTransition.Params>
+): Transition {
+    val exit = filter { it.direction == TransitionDirection.Exit }
+    val enter = filter { it.direction == TransitionDirection.Enter }
+    val transitions: MutableList<SharedElementTransitionInfo<T>> = mutableListOf()
+
+    transitionParams.forEach { transitionParam ->
+        var exitingView: View? = null
+        var enteringView: View? = null
+
+        val exitElementForId = exit.find {
```

Comment:
is the naming correct?
I don't see any id around :D

Reply:
Granted, forId is outdated naming.
But we do need the actual TransitionElement (exitElementForId) to be passed on too, not just the found View.